src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,140 @@
+                    String line = scanner.nextLine();
+                    char typeOfTask = line.charAt(2);
+                    if (typeOfTask == 'E') {
+                        String[] storedValues = line.substring(5).split("[|]", 5);
+                        boolean isDone = storedValues[0].trim().equals("X");
+                        boolean isDateOnly = storedValues[1].trim().equals("X");
+                        LocalDateTime startDate = LocalDateTime.parse(storedValues[2].trim());
+                        LocalDateTime endDate = LocalDateTime.parse(storedValues[3].trim());
+                        String description = storedValues[4].trim();
+                        taskArrayList.add(new Event(description, startDate, endDate, isDone, isDateOnly));
+                    } else if (typeOfTask == 'D') {
+                        String[] storedValues = line.substring(5).split("[|]", 4);
+                        boolean isDone = storedValues[0].trim().equals("X");
+                        boolean isDateOnly = storedValues[1].trim().equals("X");
+                        LocalDateTime date = LocalDateTime.parse(storedValues[2].trim());
+                        String description = storedValues[3].trim();
+                        taskArrayList.add(new Deadline(description, date, isDone, isDateOnly));
+                    } else {
+                        String[] storedValues = line.substring(5).split("[|]", 2);
+                        boolean isDone = storedValues[0].trim().equals("X");
+                        String description = storedValues[1].trim();
+                        taskArrayList.add(new Todo(description, isDone));
+                    }
+                }
+            } catch (FileNotFoundException e) {
+                throw new DukeException("Something went wrong: " + e.getMessage());
+
+            }
+        } else {
+            try {
+                taskList.createNewFile();
+                return new TaskList(new ArrayList<Task>());
+            } catch (IOException e) {
+                throw new DukeException("Something went wrong: " + e.getMessage());
+            }
+        }
+
+        return new TaskList(taskArrayList);
+
+    }
+
+    private void createDir() {
+
+
+        File dataDir = new File(this.filePath.substring(0, this.filePath.lastIndexOf('/')));
+        dataDir.mkdirs();
+    }
+
+
+    /**
+     * Saves the user's tasks into the hard disk. Data is stored as a formatted text file.
+     *
+     * @param tasks Tasks stored into hard disk.
+     */
+    public void save(TaskList tasks) {
+        createDir();
+        File taskListFile = new File(this.filePath);
+
+        String text = "";
+
+        for (int i = 0; i < tasks.size(); i++) {
+
+            Task currentTask = tasks.get(i);
+            if (currentTask instanceof Event) {
+                Event event = (Event) currentTask;
+                text += String.format("| E | %s | %s | %s | %s | %s\n", event.getIsDone() ? "X" : " ",

```

Comment:
I feel like it might be more aligned to OOP principles to have this String parsing done within the specific Task classes, and just calling this method here e.g. public String toFileString() { return String.format( ...); }